### PR TITLE
Hive 24471 : Add support for combiner in hash mode group aggregation

### PR DESCRIPTION
**Pull request by maheshk114**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+import static org.apache.hadoop.hive.ql.exec.Utilities.REDUCE_PLAN_NAME;
+
+// Combiner for vectorized group by operator. In case of map side aggregate, the partially
+// aggregated records are sorted based on group by key. If because of some reasons, like hash
+// table memory exceeded the limit or the first few batches of records have less ndvs, the
+// aggregation is not done, then here the aggregation can be done cheaply as the records
+// are sorted based on group by key.
+public class VectorGroupByCombiner extends MRCombiner {
+  private static final Logger LOG = LoggerFactory.getLogger(
+      VectorGroupByCombiner.class.getName());
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:63]: as above -- getName not needed

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+  void initVectorBatches() throws HiveException {
+    inputBatch = batchContext.createVectorizedRowBatch();
+
+    // Create data buffers for value bytes column vectors.
+    for (int i = firstValueColumnOffset; i < inputBatch.numCols; i++) {
+      ColumnVector colVector = inputBatch.cols[i];
+      if (colVector instanceof BytesColumnVector) {
+        BytesColumnVector bytesColumnVector = (BytesColumnVector) colVector;
+        bytesColumnVector.initBuffer();
```
[the remark this answers is not in the document]
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:204]: Actually: createVectorizedRowBatch() already calls init as the end (throught reset) any particular reason we doing this again here?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByCombiner.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+                valueTableDesc.getProperties(), null);
+
+        aggrObjectInspector = groupByOperator.getAggrObjInspector();
+        valueBuffer = new DataInputBuffer();
+        cachedValues = new Object[aggregationEvaluators.length];
+        prevKey = new DataInputBuffer();
+        valWritable = new BytesWritable();
+        prevVal = new DataInputBuffer();
+      } catch (Exception e) {
+        LOG.error(" GroupByCombiner failed", e);
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByCombiner.java:112]: log or throw

File: ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByCombiner.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  }
+
+  public static JobConf setCombinerInConf(BaseWork dest, JobConf conf, JobConf destConf) {
+    if (conf == null || !HiveConf.getBoolVar(conf, HiveConf.ConfVars.HIVE_ENABLE_COMBINER_FOR_GROUP_BY)) {
+      return conf;
+    }
+
+    // As of now, this function is called for all edges. We are interested only on edges
+    // to reducer.
+    if (!(dest instanceof ReduceWork)) {
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByCombiner.java:214]: nit. add this to the above condition

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    long numRows = 0;
+    try {
+      do {
+        numRows++;
+        writer.append(rawIter.getKey(), rawIter.getValue());
+      } while (rawIter.next());
+      combineInputRecordsCounter.increment(numRows);
+      combineOutputRecordsCounter.increment(numRows);
+    } catch(IOException e) {
+      LOG.error("Append to writer failed", e);
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:313]: Log or throw

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+  // in a vectorized row batch.
+  protected transient VectorSerializeRow<LazyBinarySerializeWrite> valueVectorSerializeRow;
+
+  // The output buffer used to serialize a value into.
+  protected transient ByteStream.Output valueOutput;
+  DataInputBuffer valueBytesWritable;
+
+  // Only required minimal configs are copied to the worker nodes. This hack (file.) is
+  // done to include these configs to be copied to the worker node.
+  protected static String confPrefixForWorker = "file.";
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:81]: not sure why we need a separate prefix for this -- the plan with the conf would already be on the workers right?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+    }
+
+    if (rw.getReducer() instanceof VectorGroupByOperator) {
+      VectorGroupByOperator vectorGroupByOperator = (VectorGroupByOperator) rw.getReducer();
+      vectorGroupByOperator.initializeOp(this.conf);
+      this.aggregators = vectorGroupByOperator.getAggregators();
+      this.aggregationBufferRow = allocateAggregationBuffer();
+      batchContext = rw.getVectorizedRowBatchCtx();
+      if ((aggregators == null) || (aggregators.length != numValueCol)) {
+        //TODO : Need to support distinct. The logic has to be changed to extract only
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:117]: Point to ticket

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+    // Create data buffers for value bytes column vectors.
+    for (int i = firstValueColumnOffset; i < inputBatch.numCols; i++) {
+      ColumnVector colVector = inputBatch.cols[i];
+      if (colVector instanceof BytesColumnVector) {
+        BytesColumnVector bytesColumnVector = (BytesColumnVector) colVector;
+        bytesColumnVector.initBuffer();
+      }
+    }
+
+    batchValueDeserializer =
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:208]: split this: to deserializeRead and vectorDeserializeRow
also: VectorizedBatchUtil.typeInfosFromStructObjectInspector(valueStructInspectors) could be reused below

**Review by pgaref — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java (hunk 5 above)
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:304]: Referring to: https://issues.apache.org/jira/browse/HIVE-25108

File: ql/src/java/org/apache/hadoop/hive/ql/exec/util/FastByteComparisons.java [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sun.misc.Unsafe;
+
+import java.lang.reflect.Field;
+import java.nio.ByteOrder;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+abstract public class FastByteComparisons {
```
[the remark this answers is not in the document]
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/util/FastByteComparisons.java:32]: Was wondering why we dont import directly either from hadoop or tez?

[the remark this answers is not in the document]
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/util/FastByteComparisons.java:32]: it's private. can not access from Hive.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java (hunk 5 above)
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java:304]: i am not sure how valid is HIVE-25141, personally i have faced issue of missing stacks and usually log the exception if its rethrown.

**Comment by pgaref** [comment]
Shall we revive this @maheshk114 ?  Seems like most of the comments have already been addressed

**Pull request closed without merging**